oap-server/server-bootstrap/src/main/resources/lal/envoy-als.yaml

```diff
@@ -0,0 +1,47 @@
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+rules:
+  - name: envoy-als
+    dsl: |
+      filter {
+        json {
+        }
+        // only collect abnormal logs (http status code >= 300, or commonProperties?.responseFlags is not empty)
+        if (parsed?.response?.responseCode < 300 && !parsed?.commonProperties?.responseFlags) {
+          abort {}
+        }
+        extractor {
+          tag 'status.code': parsed?.response?.responseCode as int
+          tag 'response.flag': parsed?.commonProperties?.responseFlags?.keySet()
+        }
+        sink {
+          sampler {
+            if (parsed?.commonProperties?.responseFlags) {
+              // use service:errorCode as sampler id so that each service:errorCode has its own sampler,
+              // e.g. checkoutservice:[upstreamConnectionFailure], checkoutservice:[upstreamRetryLimitExceeded]
+              rateLimit("${log.service}:${parsed?.commonProperties?.responseFlags?.keySet()}") {
```

Comment:
From my understanding, this means collect every response flag less than 100/s ?

Reply:
> From my understanding, this means collect less than 100/s per error response flag ?

Yes, the sampler samples the logs in dimension <service:errorResponseFlag> at most 100/s

Reply:
Got it. I have some self-doubt whether I am too paranoid to worry about OOM. Let's keep this for now. This at least makes sure we don't miss important error logs.

Reply:
> Got it. I have some self-doubt whether I am too paranoid to worry about OOM. Let's keep this for now. This at least makes sure we don't miss important error logs.

In terms of the samplers, they are rather light-weight actually, a sampler only holds a field `volatile int qps` and a field `AtomicInteger factor`.

Reply:
I know, that is why I call myself `paranoid` :P